### PR TITLE
Fix multiagent learning

### PR DESCRIPTION
**Pull request by nick-harder**
# Pull Request

## **Description**  
Learning in multi-agent setups was observed to be highly unstable, often failing to produce the desired results. This investigation focused on the update policy function in detail, identifying several critical bugs and inefficiencies. This PR addresses these issues, improving stability, performance, and consistency across the learning pipeline.

## **Major Fixes & Improvements**  
- **Fixed a critical bug** where only a single `actor_target` network was used to sample next actions.  
- **Refactored critic handling**: Moved critics and critic targets to the strategy itself, separating them from the learning role. This improves consistency and reduces potential errors.  
- **Redesigned learning strategy injection**: Learning strategies are now added to the learning role only at the end of world setup, ensuring correct order when sampling from the buffer and iterating over agents for policy updates.  
- **Single gradient step update**: Instead of looping over multiple gradient steps, a single backpropagation is now performed per update. This improves performance and efficiency, particularly on GPUs.  

## **Performance Optimizations**  
- **Batch Polyak updates**: Now applied across all agents simultaneously instead of looping through them individually, significantly improving efficiency.  
- **Optimized transition sampling**: Transitions are now sampled only once per gradient step, reducing redundant computations.  
- **More efficient state preparation**: `all_states` and `all_next_states` are now processed more efficiently, improving performance.  
- **Gradient step optimization**: Improved gradient calculation efficiency for better training speed.  
- **Gradient clipping**: Added to stabilize learning performance.  
- **Optimized tensor operations**: Converted tensors to floats when submitting bids to reduce computational overhead and improve performance.  
- **Reduced unnecessary tensor operations**: Removed unrequired `clone()` and `detach()` calls from `get_actions` and `get_observation`, reducing redundant memory operations.  
- **Direct tensor allocation on device**: Created tensors for noise and observations directly on the device instead of moving them post-creation, improving efficiency and reducing unnecessary data transfers.  

## **Architectural & Stability Improvements**  
- **Switched to AdamW optimizer**: Provides better and more stable learning performance.  
- **Improved weight initialization**: Enhances training stability for both actor and critic networks.  
- **Adjusted critic architecture**: Simplified for small agent populations while increasing complexity for larger ones.  

## **Additional Improvements & Fixes**  
- **Added historical prices to observations**: Improves agent decision-making by incorporating past price trends.  
- **Reduced foresight from 24 to 12 hours**: Adjusts the forecasting window for better balance between predictive power and stability.  
- **Removed `-1` option for gradient steps**: Eliminates confusion for users.  
- **Fixed float64 error in outputs**.  
- **Updated learning dashboard**: Now correctly calculates the average of the sum of rewards instead of the average of all rewards.  
- **Improved tensor type conversions**: Standardized tensor-to-float conversion functions to ensure consistency across computations and avoid precision-related issues.  

## Testing
Performed training in the multi-agent setup of example_02c and observed a much improved learning behavior. For large scale simulations (around 150 agents) on CPU the speed has been improved by 2x, on GPU 5x (RTX3090).

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [ ] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [x] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [x] Changes are backwards compatible, or deprecation notices added
- [ ] New dependencies added to `pyproject.toml`
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
behavior before
image

behavior now
image

### CODE REVIEW

File: assume/markets/clearing_algorithms/complex_clearing.py [hunk 1]
```diff
@@ -455,8 +455,6 @@ def clear(
 
         orderbook.sort(key=itemgetter("start_time", "end_time", "only_hours"))
```
Comment (kim-mskw) [inline assume/markets/clearing_algorithms/complex_clearing.py:456]: I double checked and we added this check as some solvers did not handel tensors well. As there are no changes in the unit operator I am assuming you did not make sure that the orders are not sending tensors to the clearing? Did you test that with all solvers?

Comment (kim-mskw) [reply assume/markets/clearing_algorithms/complex_clearing.py:456]: Nevermind found it

File: assume/strategies/learning_strategies.py [hunk 2]
```diff
@@ -34,8 +34,6 @@ def load_actor_params(self, load_path):
             obs_dim=self.obs_dim,
             act_dim=self.act_dim,
             float_type=self.float_type,
```
Comment (kim-mskw) [inline assume/strategies/learning_strategies.py:36]: Why did you delete this? It is needed for the LSTM integration?

File: assume/strategies/learning_strategies.py [hunk 3]
```diff
@@ -148,7 +144,7 @@ class RLStrategy(AbstractLearningStrategy):
     """
 
     def __init__(self, *args, **kwargs):
-        super().__init__(obs_dim=50, act_dim=2, unique_obs_dim=2, *args, **kwargs)
+        super().__init__(obs_dim=38, act_dim=2, unique_obs_dim=2, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/strategies/learning_strategies.py:147]: @kim-mskw this was your earlier suggestion from your work that 24 hours didn't really do anything and 12 is better, from your sharp values? or am I mistaken? anyway, this works also just as fine

[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/strategies/learning_strategies.py:147]: Yes true. That is why it was changed to 12 h at some point. I thought there might have been a reason why we changed it back to 24.

File: assume/reinforcement_learning/neural_network_architecture.py [hunk 4]
```diff
@@ -29,64 +29,79 @@ def __init__(
...
-        else:
-            self.FC2_1 = nn.Linear(self.obs_dim + self.act_dim, 1024, dtype=float_type)
-            self.FC2_2 = nn.Linear(1024, 512, dtype=float_type)
-            self.FC2_3 = nn.Linear(512, 128, dtype=float_type)
-            self.FC2_4 = nn.Linear(128, 1, dtype=float_type)
+            hidden_sizes = [1024, 512, 256, 128]  # Deeper network for large `n_agents`
+
+        # First Q-network (Q1)
+        self.q1_layers = self._build_q_network(hidden_sizes, float_type)
+
```
Comment (kim-mskw) [inline assume/reinforcement_learning/neural_network_architecture.py:42]: Way prettier code!

File: tests/test_learning_role.py [hunk 5]
```diff
@@ -55,11 +55,11 @@ def test_learning_init():
         assert isinstance(strategy.actor_target, Actor)
 
     # now we have a critic for every strategy
-    for str_id in learn.rl_strats.keys():
-        assert isinstance(learn.critics[str_id], CriticTD3)
-        assert isinstance(learn.target_critics[str_id], CriticTD3)
+    for strategy in learn.rl_strats.values():
+        assert isinstance(strategy.critics, CriticTD3)
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply tests/test_learning_role.py:59]: I see the benefit of having it in the startegies to ensure the proper order as higher. So I would agree to changing that.

File: assume/reinforcement_learning/algorithms/matd3.py [hunk 6]
```diff
@@ -408,99 +391,102 @@ def update_policy(self):
...
+            states, actions, next_states, rewards = (
+                transitions.observations,
+                transitions.actions,
+                transitions.next_observations,
+                transitions.rewards,
+            )
+
+            with th.no_grad():
+                # Select action according to policy and add clipped noise
+                noise = th.randn_like(actions) * self.target_policy_noise
```
Comment (kim-mskw) [inline assume/reinforcement_learning/algorithms/matd3.py:418]: Do we have different noises for the target actor and the normal actor, which ultimately results in different noise decay? I am wondering that now and checking it.

File: assume/reinforcement_learning/algorithms/matd3.py [hunk 7]
```diff
@@ -512,39 +498,97 @@ def update_policy(self):
...
-                # Compute critic loss
+                # Accumulate critic loss for this agent
                 critic_loss = sum(
                     F.mse_loss(current_q, target_Q_values)
                     for current_q in current_Q_values
                 )
+                total_critic_loss += critic_loss
+
+            # Single backward pass for all agents' critics
+            total_critic_loss.backward()
```
Comment (kim-mskw) [inline assume/reinforcement_learning/algorithms/matd3.py:509]: I am guessing you still want to use the specific critic loss for the backward pass of the critic. Are you sure this porvides the intended behavior? I find the documentation of pytorch not super conclusive

File: assume/strategies/learning_strategies.py (hunk 2 above)
Comment (kim-mskw) [reply assume/strategies/learning_strategies.py:36]: As expected, the LSTM example is not working anymore.

File: assume/reinforcement_learning/algorithms/matd3.py (hunk 6 above)
Comment (kim-mskw) [reply assume/reinforcement_learning/algorithms/matd3.py:418]: In the application of TD3 in stablebaselines§ are these noises decayed there?

**Review by kim-mskw — CHANGES_REQUESTED** [review]
- LSTM policy needs tp be fixed

Otherwise we can merge it and adjust the tensorboard.

**Comment by nick-harder** [comment]
@kim-mskw sorry for breaking the lstm policies. everything is fixed now. I have double checked the optimizer choice, we can keep it. I have also checked regarding the backpropagation, it is also correct. Gradient clipping we should keep for now as it improves the learning, but we should definitely look into it in the future

**Review by kim-mskw — APPROVED** [review]

**Comment by mthede** [comment]
Quick question on the grafana learning dashboard update @nick-harder. I noticed that the scaling changes when looking at the avg. reward of all learning units and for one specific unit:

grafik
vs. 
grafik

- Shouldn't the values be the same in a 1 learning unit example?
- In the first chart, the variable "avg_profit" confused me a bit. Is it profit or reward?
